File: assign1/scala/src/test/scala/perfect_number_test.scala
import org.scalatest.funsuite.AnyFunSuite

import org.scalatest.prop.TableDrivenPropertyChecks.forAll
import org.scalatest.prop.TableDrivenPropertyChecks.Table

trait perfect_number_test extends AnyFunSuite {

  def testIsPerfectNumber(implementation: Int => Boolean): Unit = {
    val perfectNumbers = Table(
      "number",
      6, 28
    )

    forAll(perfectNumbers) { number =>
      assert(implementation(number), s"Expected $number to be a perfect number")
    }
  }

  def testIsNotPerfectNumber(implementation: Int => Boolean): Unit = {
    val nonPerfectNumbers = Table(
      "number",
      0, 1, 5, 7, 10, 30
    )

    forAll(nonPerfectNumbers) { number =>
      assert(!implementation(number), s"Expected $number not to be a perfect number")
    }
  }
}

File: assign1/go/test/perfect_number_test.go
package test

import (
  "testing"
)

func Test_Canary(t *testing.T){
}

func test_is_perfect_number(t *testing.T, implementation func(number int) bool) {
  tests := []struct {
		number int
	}{
		{6},
		{28},
	}

	for _, tt := range tests {
    t.Run("Imperative: is perfect number", func(t *testing.T) {
			if got := implementation(tt.number); got != true {
				t.Errorf("Imperative_perfect_number(%v) = %v, want %v", tt.number, got, true)
			}
		})
	}
}

func test_is_not_perfect_number(t *testing.T, implementation func(number int) bool) {
  tests := []struct {
		number int
	}{
    {0},
    {1},
    {5},
    {7},
    {10},
    {30},
	}

	for _, tt := range tests {
    t.Run("Imperative: is not perfect number", func(t *testing.T) {
			if got := implementation(tt.number); got != false {
				t.Errorf("Imperative_perfect_number(%v) = %v, want %v", tt.number, got, false)
			}
		})
	}
}
